File: aidl.im.one/src/main/java/com/chaoya/aidlimone/MainActivity.java
package com.chaoya.aidlimone;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.chaoya.aidlim.IRemoteService;
import com.chaoya.aidlim.UserMessage;

import java.util.LinkedList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainActivity extends AppCompatActivity {


    private String TAG = this.getClass().getSimpleName();

    @BindView(R.id.content)
    EditText content;
    @BindView(R.id.send)
    TextView send;
    @BindView(R.id.list_view)
    ListView listView;

    private LinkedList<String> list = new LinkedList<>();
    private ArrayAdapter<String> adapter;
    private IRemoteService remoteService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, list);
        listView.setAdapter(adapter);

        Intent intent2 = new Intent().setComponent(new ComponentName(
                "com.chaoya.aidlimone",
                "com.chaoya.aidlimone.RemoteService"));
        bindService(intent2, mConnection2, Context.BIND_AUTO_CREATE);
    }

    private ServiceConnection mConnection1 = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.e(TAG, "onServiceConnected");
            remoteService = IRemoteService.Stub.asInterface(service);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {

        }
    };

    private ServiceConnection mConnection2 = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.e(TAG, "onServiceConnected");

            RemoteService.InnerIBinder innerIBinder = (RemoteService.InnerIBinder) service;
            RemoteService remoteService = (RemoteService) innerIBinder.getService();
            remoteService.setCallBack(new RemoteService.CallBack() {
                @Override
                public void showMessage(UserMessage message) {
                    list.addFirst("男孩说:" + message.messageContent);
                    adapter.notifyDataSetChanged();
                }
            });
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {

        }
    };

    @OnClick(R.id.send)
    public void onViewClicked() {
        if ("".equals(content.getText().toString().trim())) {
            Toast.makeText(this, "请输入信息", Toast.LENGTH_SHORT).show();
            return;
        }
        if (remoteService == null) {
            Toast.makeText(this, "服务正在启动...", Toast.LENGTH_SHORT).show();
            Intent intent1 = new Intent().setComponent(new ComponentName(
                    "com.chaoya.aidlimtwo",
                    "com.chaoya.aidlimtwo.RemoteService"));
            bindService(intent1, mConnection1, Context.BIND_AUTO_CREATE);
            return;
        }


        try {
            remoteService.sendMessage(new UserMessage(content.getText().toString().trim()));
            list.addFirst("女孩说:" + content.getText().toString().trim());
            content.setText("");
            adapter.notifyDataSetChanged();
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }
}

File: aidl.im.two/src/main/java/com/chaoya/aidlimtwo/MainActivity.kt
package com.chaoya.aidlimtwo

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.os.RemoteException
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.widget.ArrayAdapter
import android.widget.Toast
import com.chaoya.aidlim.IRemoteService
import com.chaoya.aidlim.UserMessage
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : AppCompatActivity() {

    private val TAG = this.javaClass.simpleName
    private var list: LinkedList<String> = LinkedList()
    private var adapter: ArrayAdapter<String>? = null;
    private var remoteService: IRemoteService? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setClick()

        adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, list)
        list_view?.adapter = adapter;

        val intent2 = Intent().setComponent(ComponentName(
                "com.chaoya.aidlimtwo",
                "com.chaoya.aidlimtwo.RemoteService"))
        bindService(intent2, mConnection2, Context.BIND_AUTO_CREATE)
    }

    private fun setClick() {
        send.setOnClickListener { v ->
            if ("" == content?.getText().toString().trim { it <= ' ' }) {
                Toast.makeText(this, "请输入信息", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            if (remoteService == null) {
                Toast.makeText(this, "服务正在启动...", Toast.LENGTH_SHORT).show()
                val intent1 = Intent().setComponent(ComponentName(
                        "com.chaoya.aidlimone",
                        "com.chaoya.aidlimone.RemoteService"))
                bindService(intent1, mConnection1, Context.BIND_AUTO_CREATE)
                return@setOnClickListener
            }

            try {
                remoteService?.sendMessage(UserMessage(content?.getText().toString().trim { it <= ' ' }))
                list.addFirst("男孩说:" + content?.getText().toString().trim { it <= ' ' })
                content?.setText("")
                adapter?.notifyDataSetChanged()
            } catch (e: RemoteException) {
                e.printStackTrace()
            }
         }
    }

    private val mConnection1 = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Log.e(TAG, "onServiceConnected")
            remoteService = IRemoteService.Stub.asInterface(service)
        }

        override fun onServiceDisconnected(name: ComponentName) {

        }
    }

    private val mConnection2 = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Log.e(TAG, "onServiceConnected")
            val innerIBinder = service as RemoteService.InnerIBinder
            val remoteService = innerIBinder.service as RemoteService
            remoteService.setCallBack { message ->
                list.addFirst("女孩说:" + message.messageContent)
                adapter?.notifyDataSetChanged()
            }
        }

        override fun onServiceDisconnected(name: ComponentName) {

        }
    }
}
